doc/syslog/python-logger-enhancement.md

```diff
@@ -0,0 +1,140 @@
+# SONiC Python logger enhancement #
+
+## Table of Content
+
+### Revision
+
+### Scope
+
+This document describes an enhancement to SONiC Python logger.
+
+### Definitions/Abbreviations
+
+Log identifier: The identifier of a logger instance. Log message printed by the logger instance usually contains the identifier.
+
+### Overview
+
+SONiC provides two Python logger implementations: `sonic_py_common.logger.Logger` and `sonic_py_common.syslogger.SysLogger`. Both of them do not provide the ability to change log level at real time. Sometimes, in order to get more debug information, developer has to manually change the log level in code on a running switch and restart the Python daemon. This is not convenient.
+
+SONiC also provides a C/C++ logger implementation in `sonic-platform-common.common.logger.cpp`. This C/C++ logger implementation is also a wrapper of Linux standard `syslog` which is widely used by swss/syncd. It provides the ability to set log level on fly by starting a thread to listen to CONFIG DB LOGGER table change. SONiC infrastructure also provides the Python wrapper for `sonic-platform-common.common.logger.cpp` which is `swsscommon.Logger`. However, this logger implementation also has some drawbacks:
+
+
+1. `swsscommon.Logger` assumes redis DB is ready to connect. This is a valid assumption for swss/syncd. But it is not good for a Python logger implementation because some Python script may be called before redis server starting.
+2. `swsscommon.Logger` wraps Linux syslog which only support single log identifier for a daemon. 
+
+So, `swsscommon.Logger` is not an option too.
+
+This document describes a Python logger enhancement which allows user setting log level at run time.
+
+### Requirements
+
+- Allow user to change log level of `sonic_py_common.syslogger.SysLogger` at run time.
+- Logger instance shall use default log level if redis server is not up.
+
+Note: `sonic_py_common.logger.Logger` is not in the feature scope because it is going to be deprecated.
+
+### Architecture Design
+
+`swsscommon.Logger` depends on a thread to listen to CONFIG DB LOGGER table change. It refreshes log level for each logger instances once the thread detects a DB entry change. A thread is considered heavy in a python script, especially that there are many short and simple python scripts which also use logger. To keep python logger light weight, it uses a different design than `swsscommon.Logger`:
+
+- `SysLogger` class shall be changed to a singleton.
+- `SysLogger` instance shall load log level configuration from DB during initialization stage if DB configuration is available.
+- `SysLogger` instance shall save log level configuration to DB during initialization stage if DB configuration is not available.
```

Comment:
@Junchao-Mellanox Why syslogger need to save log level to DB? The user will configure the log level which will be updated to DB and the main process would be woken up via signal to fetch the user set log level from the DB. There is no need to save log level within SysLogger class.

Reply:
we need to register logger itself to DB so that `swssloglevel -p` will print all the available log identifiers. Like this:
```
swssloglevel -p
SAI_API_ACL                   SAI_LOG_LEVEL_NOTICE
SAI_API_ARS                   SAI_LOG_LEVEL_NOTICE
SAI_API_ARS_PROFILE           SAI_LOG_LEVEL_NOTICE
SAI_API_BFD                   SAI_LOG_LEVEL_NOTICE
SAI_API_BMTOR                 SAI_LOG_LEVEL_NOTICE
SAI_API_BRIDGE                SAI_LOG_LEVEL_NOTICE
SAI_API_BUFFER                SAI_LOG_LEVEL_NOTICE
SAI_API_COUNTER               SAI_LOG_LEVEL_NOTICE
SAI_API_DASH_ACL              SAI_LOG_LEVEL_NOTICE
SAI_API_DASH_DIRECTION_LOOKUP SAI_LOG_LEVEL_NOTICE
SAI_API_DASH_ENI              SAI_LOG_LEVEL_NOTICE
SAI_API_DASH_INBOUND_ROUTING  SAI_LOG_LEVEL_NOTICE
SAI_API_DASH_METER            SAI_LOG_LEVEL_NOTICE
SAI_API_DASH_OUTBOUND_CA_TO_PASAI_LOG_LEVEL_NOTICE
SAI_API_DASH_OUTBOUND_ROUTING SAI_LOG_LEVEL_NOTICE
SAI_API_DASH_PA_VALIDATION    SAI_LOG_LEVEL_NOTICE
SAI_API_DASH_VIP              SAI_LOG_LEVEL_NOTICE
SAI_API_DASH_VNET             SAI_LOG_LEVEL_NOTICE
SAI_API_DEBUG_COUNTER         SAI_LOG_LEVEL_NOTICE
SAI_API_DTEL                  SAI_LOG_LEVEL_NOTICE
SAI_API_FDB                   SAI_LOG_LEVEL_NOTICE
SAI_API_GENERIC_PROGRAMMABLE  SAI_LOG_LEVEL_NOTICE
SAI_API_HASH                  SAI_LOG_LEVEL_NOTICE
SAI_API_HOSTIF                SAI_LOG_LEVEL_NOTICE
SAI_API_IPMC                  SAI_LOG_LEVEL_NOTICE
SAI_API_IPMC_GROUP            SAI_LOG_LEVEL_NOTICE
SAI_API_IPSEC                 SAI_LOG_LEVEL_NOTICE
SAI_API_ISOLATION_GROUP       SAI_LOG_LEVEL_NOTICE
SAI_API_L2MC                  SAI_LOG_LEVEL_NOTICE
SAI_API_L2MC_GROUP            SAI_LOG_LEVEL_NOTICE
SAI_API_LAG                   SAI_LOG_LEVEL_NOTICE
SAI_API_MACSEC                SAI_LOG_LEVEL_NOTICE
SAI_API_MCAST_FDB             SAI_LOG_LEVEL_NOTICE
SAI_API_MIRROR                SAI_LOG_LEVEL_NOTICE
SAI_API_MPLS                  SAI_LOG_LEVEL_NOTICE
SAI_API_MY_MAC                SAI_LOG_LEVEL_NOTICE
SAI_API_NAT                   SAI_LOG_LEVEL_NOTICE
SAI_API_NEIGHBOR              SAI_LOG_LEVEL_NOTICE
SAI_API_NEXT_HOP              SAI_LOG_LEVEL_NOTICE
SAI_API_NEXT_HOP_GROUP        SAI_LOG_LEVEL_NOTICE
SAI_API_POLICER               SAI_LOG_LEVEL_NOTICE
SAI_API_PORT                  SAI_LOG_LEVEL_NOTICE
SAI_API_QOS_MAP               SAI_LOG_LEVEL_NOTICE
SAI_API_QUEUE                 SAI_LOG_LEVEL_NOTICE
SAI_API_ROUTE                 SAI_LOG_LEVEL_NOTICE
SAI_API_ROUTER_INTERFACE      SAI_LOG_LEVEL_NOTICE
SAI_API_RPF_GROUP             SAI_LOG_LEVEL_NOTICE
SAI_API_SAMPLEPACKET          SAI_LOG_LEVEL_NOTICE
SAI_API_SCHEDULER             SAI_LOG_LEVEL_NOTICE
SAI_API_SCHEDULER_GROUP       SAI_LOG_LEVEL_NOTICE
SAI_API_SRV6                  SAI_LOG_LEVEL_NOTICE
SAI_API_STP                   SAI_LOG_LEVEL_NOTICE
SAI_API_SWITCH                SAI_LOG_LEVEL_NOTICE
SAI_API_SYSTEM_PORT           SAI_LOG_LEVEL_NOTICE
SAI_API_TAM                   SAI_LOG_LEVEL_NOTICE
SAI_API_TUNNEL                SAI_LOG_LEVEL_NOTICE
SAI_API_TWAMP                 SAI_LOG_LEVEL_NOTICE
SAI_API_UDF                   SAI_LOG_LEVEL_NOTICE
SAI_API_VIRTUAL_ROUTER        SAI_LOG_LEVEL_NOTICE
SAI_API_VLAN                  SAI_LOG_LEVEL_NOTICE
SAI_API_WRED                  SAI_LOG_LEVEL_NOTICE
buffermgrd                    NOTICE
coppmgrd                      NOTICE
fdbsyncd                      NOTICE
fpmsyncd                      NOTICE
gearsyncd                     NOTICE
intfmgrd                      NOTICE
nbrmgrd                       NOTICE
neighsyncd                    NOTICE
orchagent                     NOTICE
portmgrd                      NOTICE
portsyncd                     NOTICE
sflowmgrd                     NOTICE
syncd                         NOTICE
teammgrd                      NOTICE
teamsyncd                     NOTICE
tlm_teamd                     NOTICE
tunnelmgrd                    NOTICE
vlanmgrd                      NOTICE
vrfmgrd                       NOTICE
vxlanmgrd                     NOTICE
wjhcli                        NOTICE
wjhd                          NOTICE
```